### DISCUSSION TITLE
Help: U8G2 + ESP32 + SSD1327 SPI Wiring

### DISCUSSION THREAD

**Discussion by glasseffects** [olikraus/u8g2#1737] [General]
Hi, first of all, thank you for taking time to read this.

I'm new to SPI interface, mostly I had been working with I2C for it's simplicity of 4 wire connection. Due to a use case scenario, I require two Waveshare 128x128 SSD1327 OLED modules to be driven. Earlier I tried with one OLED module with I2C on Arduino Uno, however the refresh rate was causing distraction. Switching on to SPI offered drastic improvements. The pin configuration was as suggested in the demo example with the following constructor:
`U8G2_SSD1327_WS_128X128_1_4W_SW_SPI u8g2(U8G2_R0, /* clock=*/ 13, /* data=*/ 11, /* cs=*/ 10, /* dc=*/ 9, /* reset=*/ 8);`
The following PIN correlation was used:
CLK: 13
DIN: 11
CS: 10
DC: 9
RST: 8

Then I tried to port the same setup to ESP32. Following is the image of the version I'm working with:
<a href="https://ibb.co/6r5ZrJr"><img src="https://i.ibb.co/6r5ZrJr/ESP32-Pinout.jpg" alt="ESP32-Pinout" border="0"></a>
The image suggests I have two SPI peripherals. I decided to go with HSPI. However I can only recognize the following pins
HSPI CLK: D14
HSPI CS: D15
To which pins do I connect the DIN, DC and RST lines?

At present I'm running the code demo code with:
DIN: 13
CLK: 14
CS: 15
DC: 26
RST: 27
With the above configuration, the refresh rate is much slower than running with default demo code on Uno. It appears I'm not using hardware SPI interface. 
Request you to guide me.

Thanks in advance!

**Comment by olikraus** [comment]
You need to use the HW SPI constructor:
U8G2_SSD1327_WS_128X128_1_4W_HW_SPI

With the current setup, the following "may" work:
```  C++
U8G2_SSD1327_WS_128X128_1_4W_HW_SPI u8g2(U8G2_R0, /* cs=*/ 10, /* dc=*/ 9, /* reset=*/ 8, /* clock=*/ 13, /* data=*/ 11)
```

**Reply by glasseffects** [reply]
Thank you for replying.

I tried the suggested code, and I get the following compiler error:
`exit status 1`
`no matching function for call to 'U8G2_SSD1327_WS_128X128_1_4W_HW_SPI::U8G2_SSD1327_WS_128X128_1_4W_HW_SPI(const u8g2_cb_t*, int, int, int, int, int)'`

One of the example constructors in the demo, does compile , but it only specifies CS, DC and RST lines.
`U8G2_SSD1327_WS_128X128_1_4W_HW_SPI u8g2(U8G2_R0, /* cs=*/ 10, /* dc=*/ 9, /* reset=*/ 8);`
I'm not sure how to address CLK and DIN.

**Reply by glasseffects** [reply]
After some searching, finally, the following configuration worked. I hope this may be helpful for others too.
BUSY -> 4, RST -> 16, DC -> 17, CS -> SS(5), CLK -> SCK(18), DIN -> MOSI(23), GND -> GND, 3.3V -> 3.3V

I still want to thank you for the support Oliver for this wonderful library!

**Reply by olikraus** [reply]
> After some searching, finally, the following configuration worked. 

Great. Maybe you can also paste the constructor line for reference.

**Reply by glasseffects** [reply]
Sure, the following U8G2 constructor for 4 Wire SPI was used:
`U8G2_SSD1327_WS_128X128_F_4W_HW_SPI u8g2(U8G2_R0, /* cs=*/ 5, /* dc=*/ 17, /* reset=*/ 16);`

Wiring:
DIN: D23
CLK: D18
CS: D5
DC: TX2 (17)
RST: RX2 (16)

I believe, the important hardware SPI pins on ESP32 are D23 & D18. The rest are addressed through the u8g2 constructor.
